### PR TITLE
chore(metrics): Enhance comments on the internals

### LINKED ISSUE #3168: Add code comments for metrics internals
While reviewing #3094, I noticed it's hard to figure out what's going on, in particular around `event::metric::MetricValue`, mainly because the intent of the code is not clear from the code itself, and it's difficult to assert correctness. It would be great if we add doc-comments for those core structures.

P.S. Maybe it's just me, if so we can close this.

### PR DESCRIPTION
**Pull request by bruceg**
This is mostly just commentary, but contains a couple minor code tweaks.

I'd like attention in particular to the comments on the `AggregatedHistogram` and `AggregatedSummary` variants, to ensure I am representing them accurately. I was also unclear why `AggregatedSummary.values` was type `f64` and not `u32`, as my reading of the [Prometheus documentation](https://prometheus.io/docs/concepts/metric_types/#summary) seemed to indicate they are counts.

Closes #3168

### CODE REVIEW

**Review by jamtur01 — APPROVED** [review]
LGTM

**Review by lukesteensen — APPROVED** [review]

File: src/event/metric.rs [hunk 1]
```diff
@@ -18,34 +18,49 @@ pub struct Metric {
...
+    /// A Gauge represents a sampled numerical value.
+    Gauge { value: f64 },
+    /// A Set contains a set of (unordered) unique values for a key.
+    Set { values: BTreeSet<String> },
+    /// A Distribution contains a set of sampled values paired with the
+    /// rate at which they were observed.
     Distribution {
         values: Vec<f64>,
         sample_rates: Vec<u32>,
         statistic: StatisticKind,
```
Comment (MOZGIII) [inline src/event/metric.rs:45]: What does this do? Does it make sense to keep this value around?

**Comment by MOZGIII** [comment]
> I'd like attention in particular to the comments on the `AggregatedHistogram` and `AggregatedSummary` variants, to ensure I am representing them accurately. I was also unclear why `AggregatedSummary.values` was type `f64` and not `u32`, as my reading of the [Prometheus documentation](https://prometheus.io/docs/concepts/metric_types/#summary) seemed to indicate they are counts.

Looking at the proto definition of the 
https://github.com/prometheus/client_model/blob/master/metrics.proto and https://github.com/OpenObservability/OpenMetrics/blob/master/protos/prometheus.proto, it looks like they should indeed be doubles.

The reason for it is the data they represent is the value (whatever the gathered and summarised value is) per quantile (which is essentially a percentile but represented not in a percentage but as a value between 0 and 1 - i.e. 95th percentile is 0.95 quantile). Thus the quantile/value pairs might look like `(0.5, 3.12), (0.95, 1203.32)` - where `0.5` and `0.95` are percentiles/quantiles, and the `3.12` and `1203.32` are the gathered values.

What odds me a bit is our naming and representation. We probably need to redesign the metrics subsystem from the core.

Different metric types have different supported operations - for instance, we can aggregate two histograms together, but we can't aggregate two summaries. I'd suggest we implement individual types separately (`Histogram`, `Counter`, `Gauge`, `Summary`, etc) and then build the `MetricValue` from them. It should give everyone a better understanding of what's allowed and what's not at by lifting this information to the type level, rather than having a single `MetricsValue::add` that user needs to pretty much know the implementation of to use correctly.

**Review by MOZGIII — APPROVED** [review]
We probably want to do some follow-up work on refactoring the metrics.

File: src/event/metric.rs (hunk 1 above)
Comment (bruceg) [reply src/event/metric.rs:45]: The `statistic` field comes from the statsd source, and is used by many of the metrics sinks. I don't think we can just drop it.

**Pull request merged**